compiler/dyno/lib/framework/ErrorBase.cpp

```diff
@@ -339,33 +366,243 @@ void ErrorAmbiguousConfigSet::write(ErrorWriterBase& wr) const {
   auto& name1 = std::get<1>(info);
   auto& name2 = std::get<2>(info);
   auto variable = std::get<const uast::Variable*>(info);
-  wr.heading(kind_, locationOnly(variable),
+  wr.heading(kind_, type_, locationOnly(variable),
             "config set ambiguously via '-s", name1, "' and '-s", name2, "'");
 }
 
 void ErrorImplicitSubModule::write(ErrorWriterBase& wr) const {
   const char* stmtKind = "require', 'use', and/or 'import";
   auto mod = std::get<const uast::Module*>(info);
+  auto requireUseOrImport = std::get<const uast::AstNode*>(info);
   auto path = std::get<UniqueString>(info);
-  wr.heading(kind_, mod,
-             "as written, '", mod->name().c_str(), "' is a sub-module of the "
-             "module created for file '", path.c_str(), "' due to the "
-             "file-level '", stmtKind, "' statements.  If you meant for '",
-             mod->name().c_str(), "' to be a top-level module, move the '",
-             stmtKind, "' statements into its scope.");
+  wr.heading(kind_, type_, mod, "an implicit module for file '", path, "' is "
+            "being created with '", mod->name(), "' as a sub-module");
```

Comment:
To my tastes, the use of "file" in this error message, while very precise, also isn't strictly necessary.  I.e., I'll understand 'foo.chpl' to be a filename without spending a word on it.  That's not a reason to change it necessarily, but if my goal were brevity, I probably would.

It's not obvious to me why we use different brief messages for `ErrorImplicitSubModule` vs. `implicitModule`—in both cases a sub-module is being created, and in both cases, it's the presence of file-scope code that is causing a file-scope module to be created and any declared modules to become sub-modules.  So should they / could they be combined into a single error?  My guess here is that we started out by implementing one in the compiler (probably this sub-module one) because of the common/confusing case for C programmers who were trying to put only non-executable code like `use` or `require` or `import` into their file scope like a `#include`.  And then later we added the other error for other file-scope code?

Maybe put a different way, once I have a mechanism for squashing a specific class of warnings/errors, it's not clear why I'd need to use two different error classes to squash this family of warnings rather than just one since they seem to be caused by, and explaining, the same thing.  And, as a user, it might not be clear to me why I get different error messages for the two cases.

If I were to experiment with this, I'd remove this error and see whether the other catches it and is sufficiently clear on its own.

Reply:
The error looks like this:
```
─── warning in errors/implicitSubModule.chpl:4 [ImplicitFileModule] ───
  An implicit module named 'implicitSubModule' is being introduced to contain this file-scope code
  The following is the offending file-scope statement:
      |
    4 | use List;
      |
  The implicit module 'implicitSubModule' is being created because the above code is outside of any module declarations (e.g., 'module M').
  Note that all of the file's contents will be placed into the new 'implicitSubModule' module.
```
Or, in brief mode,
```
warning in errors/implicitSubModule.chpl:4: an implicit module named 'implicitSubModule' is being introduced to contain this file-scope code
```

Reply:
To me, that seems good enough that I'd merge it into one error class rather than keeping it as two similar-but-distinct ones.

Was the word "offending" there the last time I looked at it?  [checking a snapshot of the webpage, I think not?)]  To me that has strong and overly negative connotations, suggesting that the code is "wrong".  It isn't, it's just a potential case of a new user being confused.  Something like "the first file-scope statement" seems more value-neutral to me.

W.r.t. the lack of a period on the first sentence of the full description mode:  It seems odd to me to have a period after the other sentences, but not this one...  Did this relate to the point you were making about some brief mode errors being complete sentences and others not?

Reply:
How are you dealing with the change in capitalization in full mode vs. lowercase in brief mode?  Guessing, I could imagine doing a similar thing for punctuation in which the error message would embed the punctuation (say) and then have it be removed in brief mode.

Reply:
I am specifically capitalizing errors in detailed error messages, and leaving them "as is" otherwise (some brief mode messages are capitalized, and I didn't want to break any more .good files than necessary). I can definitely write code to strip terminating punctuation in brief mode.

Reply:
Oh, so the same string is not used for the brief message as for the first line of the detailed one?  Something made me think that it was, so the capitalization differences above suggested heuristic post-processing to me:

> An implicit module named 'implicitSubModule' is being introduced to contain this file-scope code
vs.
> an implicit module named 'implicitSubModule' is being introduced to contain this file-scope code

If they are different strings, can we just tack a period onto the detailed version's string?

Reply:
They are the same string, and you are completely correct that post-processing is going on. I have added an additional post-processing branch that strips punctuation; the changed `.good` files reflect this change (in the message being discussed, as well as in some other messages that previously had periods)